File: sdks/dart/lib/src/crypto/bytes.dart
// Byte/encoding helpers shared by the chat crypto layer. Dart's dart:convert
// and dart:typed_data already give us a real cross-platform base64 and
// random source, so this file is much thinner than its JS counterpart (which
// has to paper over browser/Node/React Native differences by hand).
import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';

String toBase64(List<int> bytes) => base64.encode(bytes);

Uint8List fromBase64(String encoded) => base64.decode(encoded);

Uint8List utf8ToBytes(String s) => Uint8List.fromList(utf8.encode(s));

String bytesToUtf8(List<int> bytes) => utf8.decode(bytes);

/// Concatenates any number of byte lists into one Uint8List.
Uint8List concatBytes(List<List<int>> parts) {
  final total = parts.fold<int>(0, (n, p) => n + p.length);
  final out = Uint8List(total);
  var offset = 0;
  for (final p in parts) {
    out.setRange(offset, offset + p.length, p);
    offset += p.length;
  }
  return out;
}

final _secureRandom = Random.secure();

/// Cryptographically random bytes.
Uint8List randomBytes(int length) {
  final out = Uint8List(length);
  for (var i = 0; i < length; i++) {
    out[i] = _secureRandom.nextInt(256);
  }
  return out;
}

/// Encodes a non-negative integer as 4 big-endian bytes — used to bind a
/// counter (e.g. a ratchet chain/message index) into associated data.
Uint8List u32be(int n) {
  final out = Uint8List(4);
  final view = ByteData.view(out.buffer);
  view.setUint32(0, n, Endian.big);
  return out;
}

/// Constant-structure lexicographic comparison of two byte lists — used to
/// pick a side-independent canonical order (e.g. for associated data), not
/// for anything requiring timing-safety.
int compareBytes(List<int> a, List<int> b) {
  final len = a.length < b.length ? a.length : b.length;
  for (var i = 0; i < len; i++) {
    if (a[i] != b[i]) return a[i] - b[i];
  }
  return a.length - b.length;
}

File: sdks/js/src/crypto/bytes.ts
// Cross-environment byte/encoding helpers. The chat crypto layer runs in
// browsers, Node, and React Native — none of which reliably share a single
// base64 API — so every conversion goes through here rather than assuming
// `Buffer` or `btoa`/`atob` exist.

/** Encodes bytes as base64, using Buffer when available (Node) and the
 * browser's btoa otherwise. */
export function toBase64(bytes: Uint8Array): string {
  if (typeof Buffer !== 'undefined') {
    return Buffer.from(bytes).toString('base64');
  }
  let binary = '';
  for (let i = 0; i < bytes.length; i++) binary += String.fromCharCode(bytes[i]);
  // eslint-disable-next-line no-undef
  return btoa(binary);
}

/** Decodes a base64 string produced by toBase64. */
export function fromBase64(b64: string): Uint8Array {
  if (typeof Buffer !== 'undefined') {
    return new Uint8Array(Buffer.from(b64, 'base64'));
  }
  // eslint-disable-next-line no-undef
  const binary = atob(b64);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) bytes[i] = binary.charCodeAt(i);
  return bytes;
}

const textEncoder = new TextEncoder();
const textDecoder = new TextDecoder();

export function utf8ToBytes(s: string): Uint8Array {
  return textEncoder.encode(s);
}

export function bytesToUtf8(b: Uint8Array): string {
  return textDecoder.decode(b);
}

/** Concatenates any number of byte arrays into one. */
export function concatBytes(...parts: Uint8Array[]): Uint8Array {
  const total = parts.reduce((n, p) => n + p.length, 0);
  const out = new Uint8Array(total);
  let offset = 0;
  for (const p of parts) {
    out.set(p, offset);
    offset += p.length;
  }
  return out;
}

/** Cryptographically random bytes, cross-environment. */
export function randomBytes(length: number): Uint8Array {
  const out = new Uint8Array(length);
  globalThis.crypto.getRandomValues(out);
  return out;
}

/** Encodes a non-negative integer as a fixed 4-byte big-endian Uint8Array —
 * used to bind a counter (e.g. a chain/message index) into a KDF input. */
export function u32be(n: number): Uint8Array {
  const out = new Uint8Array(4);
  new DataView(out.buffer).setUint32(0, n, false);
  return out;
}
